tools/istio-iptables/pkg/validation/validator.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"time"
+
+	"istio.io/istio/tools/istio-iptables/pkg/config"
+	"istio.io/istio/tools/istio-iptables/pkg/constants"
+)
+
+type Validator struct {
+	Config *Config
+}
+
+type Config struct {
+	ServerListenAddress string
+	ServerOriginalPort  uint16
+	ServerOriginalIP    net.IP
+}
+type Service struct {
+	Config *Config
+}
+type Client struct {
+	Config *Config
+}
+
+func (validator *Validator) Run() error {
+	s := Service{
+		validator.Config,
+	}
+	sError := make(chan error, 1)
+	sTimeout := time.After(30 * time.Second)
+	go func() {
+		sError <- s.Run()
+	}()
+
+	// infinite loop
+	go func() {
+		c := Client{Config: validator.Config}
+		for {
+			_ = c.Run()
+			// Avoid spamming the request to the validation server.
+			// Since the TIMEWAIT socket is cleaned up in 60 second,
+			// it's maintaining 60 TIMEWAIT sockets. Not big deal.
+			time.Sleep(time.Second)
+		}
+	}()
+	select {
+	case <-sTimeout:
+		return errors.New("validation timeout")
+	case err := <-sError:
+		if err == nil {
+			fmt.Println("validation passed!")
+		} else {
+			fmt.Println("validation failed:" + err.Error())
+		}
+		return err
+	}
+}
+
+func NewValidator(config *config.Config) *Validator {
+	return &Validator{
+		Config: &Config{
+			ServerListenAddress: ":" + config.InboundCapturePort,
```

Comment:
This is listening on all IP addresses, including the pod's routable address. This means that this may receive connections from outside of the pod, which is not acceptable. This socket must only listen on the local interface. So you must specify the host address parameter, e.g., for IPv4:
```suggestion
			ServerListenAddress: "127.0.0.1:" + ...,
```
For IPv6:
```suggestion
			ServerListenAddress: "[::1]:" + ...,
```

Reply:
> This is listening on all IP addresses, including the pod's routable address. 

Yes. To keep the code simple

> This means that this may receive connections from outside of the pod, which is not acceptable

Why it is not acceptable? Those connection see RST anyway, if we don't listen here.

Reply:
That is indeed a behavior change though

Reply:
I'm OK with going with IPv4 only, but you need at least to listen on `127.0.0.1`.

Reply:
done